myia/analyze/value.py

```diff
@@ -0,0 +1,131 @@
+"""Value inference."""
+from typing import Dict, Callable
+from .graph import Plugin
+
+from myia import primops as P
+from myia.primops import Primitive
+from myia.utils import Named
+from myia.anf_ir import Constant, Apply, Parameter, Graph
+from myia.py_implementations import implementations as py_implementations
+
+ESTIMATORS: Dict[Primitive, Callable] = dict()
+
+
+NO_VALUE = Named('NO_VALUE')
+NOT_CONSTANT = Named('NOT_CONSTANT')
+
+
+def register_estimator(prim: Primitive):
+    """Register an estimator for a primitive.
+
+    This serves to estimate the value of a primitive in the present of
+    incomplete information (not all arguments are constants).  In some
+    cases we can know the value (like mul(0, x) -> 0).  If you
+    can't do anything unless you know all the arguments don't
+    implement this.
+    """
+    def deco(e: Callable):
+        assert prim not in ESTIMATORS
+        ESTIMATORS[prim] = e
+        return e
+    return deco
+
+
+# SCCP
+class ValuePlugin(Plugin):
+    """Implementation of SCCP (Sparse Conditional Constant Propagation).
+
+    This has some trouble with loops and such since those are
+    implemented using recursive calls and we don't handle that.
+
+    It does deal with conditionals ok.
+    """
+
+    NAME = "value"
+
+    def __init__(self, *, implementations: Dict[Primitive, Callable] = None,
+                 estimators: Dict[Primitive, Callable] = None) -> None:
+        """Create a ValuePlugin."""
+        if implementations is None:
+            implementations = py_implementations
+        self._implementations = implementations
+        if estimators is None:
+            estimators = ESTIMATORS
+        self._estimators = estimators
+
+    def visit(self, fn: Callable, value):
+        """Nothing to visit."""
+        raise self.analyzer.DU.VisitError  # pragma: no cover
+
+    def on_attach(self):
+        """Add shortcuts."""
+        # Don't explicitely depend on TypePlugin since that will create a cycle
```

Comment:
*Does* it require a `TypePlugin` to be present at all, though?

Reply:
Not at the moment, but probably a bit later it will.  I could make it a soft dependency though.